Extend framework to support Counter-timer Kernel Control register access from user space

sachin-saxena-nxp opened the pull request:
While doing performance bench-marking using PMU, an issue of "Application cpu cycle spikes in timeboud 5G applications running on ARM" is seen. Context switching is the main reason for the spikes in cycle count when measured with PMU.

I have found out a solution for achieving hard real time behavior of a DPDK application on ARM by playing with core settings and by masking/unmasking "ARM per-core architected timer interrupts" which avoid these unwanted context switching on an isolated CPU.

Intention is that any DPDK or general application can enable this access via this kernel module and then mask/unmask arch_timer irq, using ARM "cntp_ctl_el0" register, around the Critical time bound processing piece of code.

Reference:
https://developer.arm.com/documentation/ddi0601/2020-12/AArch64-Registers/CNTKCTL-EL1--Counter-timer-Kernel-Control-register

jerinjacobk commented on pmu_el0_cycle_counter.c:
```diff
@@ -99,6 +104,28 @@ disable_cycle_counter_el0(void* data)
 
 }
 
+static void
```
move to pmu_el0_timer_control.c

jerinjacobk commented on pmu_el0_cycle_counter.c:
```diff
@@ -99,6 +104,28 @@ disable_cycle_counter_el0(void* data)
+}
+
 static ssize_t
 pmccntr_show(char *arg, size_t size)
```
change the name to support new IOCTLs

jerinjacobk commented on pmu_el0_cycle_counter.c:
```diff
@@ -243,6 +270,18 @@ pmuctl_ioctl(struct file *f, unsigned int cmd, unsigned long arg)
 			on_each_cpu(disable_cycle_counter_el0, NULL, 1);
 		break;
+
+	case PMU_IOC_CNTKCTL: /* Enable Physical Timer Control EL0 access */
```
Move to pmu_el0_timer_control.c and call from here.

sachin-saxena-nxp replied:
Just to confirm... your suggestion is to move all below handling in pmu_el0_timer_control.c  and callthe related function from this IOCTL handler. 

Please confirm.

jerinjacobk replied:
Yes

jerinjacobk commented:
> This approach and its side effect has been discussed with ARM too.

Please share the public ref to same if available in the commit log.

sachin-saxena-nxp commented:
> > This approach and its side effect has been discussed with ARM too.
> 
> Please share the public ref to same if available in the commit log.

The discussion is over NXP & ARM internal mailchain on behalf of an customer issue. I am not sure whether i can share that or not. But, I was discussing this matter recently with  "Honnappa Nagarahalli <Honnappa.Nagarahalli@arm.com>" & "Ruifeng Wang <Ruifeng.Wang@arm.com>".

We are referring to ARM developer guide for all the information  @ https://developer.arm.com/documentation/ddi0601/2020-12/AArch64-Registers/CNTKCTL-EL1--Counter-timer-Kernel-Control-register

jerinjacobk commented:
>  I am not sure whether i can share that or not

No need to share then.  Just remove "This approach and its side effect has been discussed with ARM too." from git commit log.
And just keep https://developer.arm.com/documentation/ddi0601/2020-12/AArch64-Registers/CNTKCTL-EL1--Counter-timer-Kernel-Control-register

Commit:
Extend framework to support Counter-timer Kernel Control
register access from user space

 - ARM core architecture timer may interfere when doing performance
   bench-marking using PMU. The arch-timer interrupt caused context
   switching which is the main reason for the spikes in user space
   cycle count when measured with PMU. One can mask/unmask "ARM per
   core architected timer interrupts" which avoid these unwanted context
   switching on an isolated CPU.
 - The new interface can be used by any DPDK or general application to
   mask/unmask arch_timer irq, using ARM "cntp_ctl_el0" register, around
   the Critical time bound processing piece of code. However note that this
   should be used carefully as no scheduling will be available in this mode.

   Reference:
   https://developer.arm.com/documentation/ddi0601/2020-12/AArch64-Registers/CNTKCTL-EL1--Counter-timer-Kernel-Control-register

Signed-off-by: Sachin Saxena <sachin.saxena@oss.nxp.com>
Reviewed-by:   Hemant Agrawal <hemant.agrawal@nxp.com>

sachin-saxena-nxp commented:
I have to rename  pmu_el0_cycle_counter.c -> armv8_pmu_el0_cycle_counter.c as in Makefile module name cannot be same as filename when including several files.
